### DISCUSSION TITLE
Separators on Toolbar

### DISCUSSION THREAD

**Discussion by varlesh** [tsujan/Kvantum#578] [Q&A]
It's possible themable & draw separators on Toolbar Lineedit & Grouped Toolbar Buttons (not have yet)?
I create id `tbutton-separator` as `spin-separator` states but this doesn't worked...
```
[ToolbarLineEdit]
inherits=PanelButtonCommand
frame.element=tbutton
interior.element=tbutton
indicator.element=tbutton-separator
text.margin.left=1
text.margin.right=1
```
![sep](https://user-images.githubusercontent.com/8083855/102448575-e1cd8700-404b-11eb-9a45-24c974fe8bab.png)

**Comment by varlesh** [comment]
on `ToolbarLineEdit` anyway draw `spin-separator`

**Reply by tsujan** [reply]
For spinbox, read `Theme-Making.pdf` → `Indicators` → `IndicatorSpinBox` for a thorough explanation. You can also take a look at `KvGnome` or `KvYaru`.

As for toolbar buttons, if they are grouped (`group_toolbar_buttons=true`), a separator can be drawn between adjacent buttons if one of them is toggled. Read `Theme-Making.pdf` → `Grouped Toolbar Buttons and Separators` for a complete explanation. You can also take a look at `KvCurvesLight`.

For toolbar separators, read `Theme-Making.pdf` → `Indicators` → `Toolbar`, but I don't think you meant them; they aren't related to toolbar buttons and no theme that comes with Kvantum has them because, to me, they aren't elegant.

All in all, you'll encounter no problem if you exactly follow the documentation.

And thanks for opening a "discussion" instead of an "issue"!

**Reply by tsujan** [reply]
BTW, I might not be near my computer until tonight. If you couldn't make separators for your theme, attach it as a zipped file; I'll fix your theme as soon as I find the time.

**Comment by varlesh** [comment]
Thx, i already read this manual. 
But only **Toggled** Toolbar Button use separator, **Normal** & **Focused** states not possible used (((
![image](https://user-images.githubusercontent.com/8083855/102456964-6c69b280-405b-11eb-902e-bc1f3fc638ef.png)

About **IndicatorSpinBox** 
I can't draw different style `LineEdit` & `ToolbarLineEdit`, because `ToolbarLineEdit` not have his separator and use separator from `LineEdit` - it's limits opportunities

**Reply by tsujan** [reply]
> But only Toggled Toolbar Button use separator, Normal & Focused states not possible used 

Yes, the documentation tells the reason clearly. the separator is drawn only because the toggled button may have a sunken shape.

Each extra element takes some CPU time to be found and drawn. If I didn't limit the number of extra elements, Kvantum wouldn't be as light as it is. The goal has been a versatile style engine, not a bloated tool for cloning GTK themes.

> About IndicatorSpinBox
> I can't draw different style LineEdit & ToolbarLineEdit,

IndicatorSpinBox has nothing to do with the separator of LineEdit or ToolbarLineEdit. Everything can be found in the documentation.

**Comment by varlesh** [comment]
Separator on grouped **ToolbarButton**:
- Normal (not have)
- Focused (not have)
- Toggled
- Pressed (not have)

Separators (IndicatorSpinBox) on **LineEdit**
- Normal
- Focused
- Toggled
- Pressed

Separators (IndicatorSpinBox) on **ToolbarLineEdit**
- Normal (not have)
- Focused (not have)
- Toggled (not have)
- Pressed (not have)

**Comment by varlesh** [comment]
Not critical for `IndicatorSpinBox`, but will be cool added separators for grouped **Buttons**:
Normal
![b1](https://user-images.githubusercontent.com/8083855/102465414-0aaf4580-4067-11eb-9bf8-d5b3d46efea1.png)

Focused
![b2](https://user-images.githubusercontent.com/8083855/102465430-11d65380-4067-11eb-8401-cf2becc5ca29.png)

Pressed
![b3](https://user-images.githubusercontent.com/8083855/102465455-1864cb00-4067-11eb-9acb-2661ab71ed4c.png)

**Reply by tsujan** [reply]
As explained in https://github.com/tsujan/Kvantum/discussions/578#discussioncomment-218458, extra elements have a cost and I've limited that cost is a reasonable way.

**Comment by varlesh** [comment]
> Yes, the documentation tells the reason clearly. the separator is drawn only because the toggled button may have a sunken shape.

Not necessary. How you see, it's can be draw on other states. And spin have this states separators but buttons not...

> Each extra element takes some CPU time to be found and drawn. If I didn't limit the number of extra elements, Kvantum wouldn't be as light as it is. The goal has been a versatile style engine, not a bloated tool for cloning GTK themes.

But our world is not perfect and toolkits too. It's old war - GTK & Qt. I'm port GTK Theme to Kvantum, because like uniform style. Some apps not have alternate in Qt and many users are forced to use GTK-apps on Qt-desktop. And i don't know why GTK Themes developers don't want port Qt-style or themes to GTK. 
Anyway, Kvantum already have some popular GTK Themes and i think 3 separators - it's not take more power or time loading on CPU.

**Reply by tsujan** [reply]
> Not necessary...

Spinbox buttons are always the same; whether they're on toolbars or not. The spinbox separator is drawn on its left button, not on its line-edit. Therefore, it's always the same, whether on toolbars or elsewhere.

Only tool-buttons, combo-boxes and line-edits have toolbar variants. Those (optional) variants exist in no other Qt style because Qt doesn't support them. *I've enforced them in Kvantum.* If they were easy to add, all styles would have them. (The same is true for light-and-dark themes.)

> But our world is not perfect and toolkits too.

In all of my apps, I pay utmost attention to the CPU time and never sacrifice it for needless properties.

> ... it's not take more power or time loading on CPU.

It's not that simple at all. You might want to search for "-separator" in more than 16000 code lines of `Kvantum.cpp` to see that.

Last but not least, as I've said repeatedly here and there, Kvantum wasn't created to clone GTK themes. It can do so in a way no other style can but that's just a byproduct, not a goal (see the last line of `INSTALL.md`).
